src/passes/OptimizeInstructions.cpp

```diff
@@ -3994,26 +4010,40 @@ struct OptimizeInstructions
         Binary* add;
         Const* c1;
         Const* c2;
-        if ((matches(curr,
-                     binary(binary(&add, Add, any(), ival(&c1)), ival(&c2))) ||
-             matches(curr,
-                     binary(binary(&add, Add, any(), ival(&c1)), ival(&c2)))) &&
+        if (matches(curr,
+                    binary(binary(&add, Add, any(), ival(&c1)), ival(&c2))) &&
             !canOverflow(add)) {
-          if (c2->value.geU(c1->value).getInteger()) {
-            // This is the first line above, we turn into x > (C2-C1)
-            c2->value = c2->value.sub(c1->value);
-            curr->left = add->left;
-            return curr;
+          // Also check for an overflow when doing C2-C1. We subtract the
+          // smaller from the larger, which avoids unsigned overflows, but if
+          // the comparison is signed we need to also consider the sign bit.
+          // E.g. 0x80000000 - 1 = 0x7fffffff is fine as unsigned, but as a
+          // signed number the -1 turns a negative number into a positive one.
+          bool constsOverflow = false;
+          if (isSignedOp(curr->op)) {
+            auto typeMaxBits = getBitsForType(add->type);
+            auto c1Bits = Bits::getMaxBits(c1, this);
+            auto c2Bits = Bits::getMaxBits(c2, this);
+            if (c1Bits + c2Bits >= typeMaxBits) {
```

Comment:
Why is this the correct condition to check? It seems that problems only arise when the greater unsigned C can be at least INT_MIN and the subtraction would take it below INT_MIN, i.e. its signed interpretation undeflows from negative to positive. Notably, if the greater unsigned C starts out less than INT_MIN, there can never be a problem since we know that the subtraction never wraps across zero.

Actually, since C1 and C2 are known constants, we can just directly check whether the subtraction would wrap across INT_MIN and don't need to generalize to consider max bits at all.

And if the subtraction C1 - C2 would wrap across INT_MIN, then we know that C1 is negative and C2 is positive, so we could have taken C2 to be our larger value and used the C2 - C1 code path instead.

We know the constant values and we know whether to interpret them as signed or unsigned, so we should always be able to choose the substraction that avoids the kind of underflow we care about avoiding.

Reply:
Hmm, I'm not sure. If we always subtract the smaller from the larger (using the proper sign, and not the unsigned sign as we did before this PR), then given `0, 0x80000000` then for the signed case we'd be doing `0 - 0x80000000` which overflows for another reason: `0x80000000` is the weird nonzero value for whom `x == -x = 0 - x`; and in particular `0 - 0x7fffffff` is positive while `0 - 0x80000000` is negative, so this is nonlinear (I'm not sure if that's what you mean by "wraps across INT_MIN" or not?). And, specifically for the rule here,
```
;; this is always false
x + 0x80000000 > 0
;; apply the rule
x > 0 - 0x80000000 = 0x80000000
;; this is sometimes true, so we broke something
```

I spent some time thinking on this but just don't think it's worth trying to handle all these cases given the complexity and low benefit. Unless I'm missing a simple solution? Otherwise, I think it's fine if we don't optimize all the things; just handling the common cases is enough, which I think this PR does.

Reply:
Sure, 0x80000000 is an edge case to consider. My high level point is that we have the constant values, so doing exhaustive case analysis is feasible and wouldn't be that complicated.

But being more conservative sounds fine as well. In that case, where does the idea to compare `c1Bits + c2Bits >= typeMaxBits` come from? It seems correct, but also somewhat arbitrary and not much simpler or obviously correct than the precise analysis.

Reply:
Another conservative approach would be to keep the existing logic but restrict the optimization to only work on unsigned comparisons.

Reply:
My idea was that if the total sum of bits can't even add up to the overflow area, then we're ok. But actually I guess that's wrong, as any signed number has max bits of the maximum...

I think it would be nice to also get small signed values here. If I can't figure out a good way though then I agree focusing on unsigned ones is enough.